File: c_programs_and_containers/build_blocks/ros_planner_ii_robot/libs/mqtt_pubsub.c
/**
 * @file mqtt_pubsub.c
 * @brief Streaming MQTT pub/sub with thread-safe ring buffer.
 *
 * mosquitto's on_message callback fires on its network thread.
 * We buffer messages in a lock-free(ish) ring and let the main
 * thread drain them via mqps_poll() / mqps_drain().
 *
 * Copied from knowledge_base/mqtt/ for standalone robot deployment.
 */

#include "mqtt_pubsub.h"
#include <mosquitto.h>
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdatomic.h>

/* ================================================================== */
/*  Ring buffer (single producer / single consumer with mutex)         */
/* ================================================================== */

typedef struct {
    MqpsMsgSlot  slots[MQPS_RING_SIZE];
    atomic_uint  head;   /* next write position (producer: callback thread) */
    atomic_uint  tail;   /* next read position  (consumer: main thread)    */
    atomic_uint  dropped;
    pthread_mutex_t lock;
} MqpsRing;

static void ring_init(MqpsRing *r) {
    memset(r->slots, 0, sizeof(r->slots));
    atomic_store(&r->head, 0);
    atomic_store(&r->tail, 0);
    atomic_store(&r->dropped, 0);
    pthread_mutex_init(&r->lock, NULL);
}

static void ring_destroy(MqpsRing *r) {
    pthread_mutex_destroy(&r->lock);
}

static uint32_t ring_used(MqpsRing *r) {
    uint32_t h = atomic_load(&r->head);
    uint32_t t = atomic_load(&r->tail);
    return (h - t) & (MQPS_RING_SIZE - 1);
}

/* Push a message (called from mosquitto callback thread) */
static bool ring_push(MqpsRing *r, const char *topic,
                       const void *payload, int payload_len,
                       int qos, bool retain) {
    pthread_mutex_lock(&r->lock);

    uint32_t h = atomic_load(&r->head);
    uint32_t t = atomic_load(&r->tail);
    uint32_t next = (h + 1) & (MQPS_RING_SIZE - 1);

    if (next == t) {
        /* Ring full -- drop oldest */
        atomic_fetch_add(&r->dropped, 1);
        atomic_store(&r->tail, (t + 1) & (MQPS_RING_SIZE - 1));
    }

    MqpsMsgSlot *slot = &r->slots[h & (MQPS_RING_SIZE - 1)];

    /* Copy topic */
    if (topic) {
        strncpy(slot->topic, topic, MQPS_MAX_TOPIC - 1);
        slot->topic[MQPS_MAX_TOPIC - 1] = '\0';
    } else {
        slot->topic[0] = '\0';
    }

    /* Copy payload */
    int copy_len = payload_len;
    if (copy_len > MQPS_MAX_PAYLOAD - 1) copy_len = MQPS_MAX_PAYLOAD - 1;
    if (payload && copy_len > 0) {
        memcpy(slot->payload, payload, copy_len);
    }
    slot->payload[copy_len] = '\0';
    slot->payload_len = (uint32_t)copy_len;
    slot->qos = qos;
    slot->retain = retain;
    slot->valid = true;

    atomic_store(&r->head, next);
    pthread_mutex_unlock(&r->lock);
    return true;
}

/* Pop up to max messages (called from main thread) */
static int ring_pop(MqpsRing *r, MqpsMsgSlot *out, int max) {
    pthread_mutex_lock(&r->lock);

    int count = 0;
    while (count < max) {
        uint32_t h = atomic_load(&r->head);
        uint32_t t = atomic_load(&r->tail);
        if (h == t) break;  /* empty */

        MqpsMsgSlot *slot = &r->slots[t & (MQPS_RING_SIZE - 1)];
        if (slot->valid) {
            memcpy(&out[count], slot, sizeof(MqpsMsgSlot));
            slot->valid = false;
            count++;
        }
        atomic_store(&r->tail, (t + 1) & (MQPS_RING_SIZE - 1));
    }

    pthread_mutex_unlock(&r->lock);
    return count;
}

/* ================================================================== */
/*  MqttPubSubStream handle                                            */
/* ================================================================== */

struct MqttPubSubStream {
    struct mosquitto *mosq;
    char   host[256];
    int    port;
    char   client_id[128];
    bool   connected;
    MqpsRing ring;
    uint64_t published;
    uint64_t received;
    uint64_t polled;
};

/* ================================================================== */
/*  mosquitto callbacks                                                */
/* ================================================================== */

static void on_connect(struct mosquitto *mosq, void *userdata, int rc) {
    (void)mosq;
    MqttPubSubStream *ps = (MqttPubSubStream *)userdata;
    if (rc == 0) {
        ps->connected = true;
    } else {
        fprintf(stderr, "[mqps] connect failed: rc=%d\n", rc);
        ps->connected = false;
    }
}

static void on_disconnect(struct mosquitto *mosq, void *userdata, int rc) {
    (void)mosq;
    MqttPubSubStream *ps = (MqttPubSubStream *)userdata;
    ps->connected = false;
    if (rc != 0) {
        fprintf(stderr, "[mqps] unexpected disconnect: rc=%d\n", rc);
    }
}

static void on_message(struct mosquitto *mosq, void *userdata,
                        const struct mosquitto_message *msg) {
    (void)mosq;
    MqttPubSubStream *ps = (MqttPubSubStream *)userdata;
    ps->received++;
    ring_push(&ps->ring, msg->topic,
              msg->payload, msg->payloadlen,
              msg->qos, msg->retain);
}

/* ================================================================== */
/*  Public API                                                         */
/* ================================================================== */

MqttPubSubStream *mqps_create(const char *host, int port,
                               const char *client_id) {
    MqttPubSubStream *ps = calloc(1, sizeof(MqttPubSubStream));
    if (!ps) return NULL;

    strncpy(ps->host, host ? host : "localhost", sizeof(ps->host) - 1);
    ps->port = port > 0 ? port : 1883;
    strncpy(ps->client_id, client_id ? client_id : "mqps_client",
            sizeof(ps->client_id) - 1);

    ring_init(&ps->ring);

    mosquitto_lib_init();

    ps->mosq = mosquitto_new(ps->client_id, false, ps);  /* clean_session=false */
    if (!ps->mosq) {
        free(ps);
        return NULL;
    }

    mosquitto_connect_callback_set(ps->mosq, on_connect);
    mosquitto_disconnect_callback_set(ps->mosq, on_disconnect);
    mosquitto_message_callback_set(ps->mosq, on_message);

    return ps;
}

void mqps_destroy(MqttPubSubStream *ps) {
    if (!ps) return;
    if (ps->connected) {
        mosquitto_disconnect(ps->mosq);
    }
    mosquitto_destroy(ps->mosq);
    ring_destroy(&ps->ring);
    free(ps);
}

bool mqps_connect(MqttPubSubStream *ps, int timeout_ms) {
    if (!ps || !ps->mosq) return false;

    int rc = mosquitto_connect(ps->mosq, ps->host, ps->port, 60);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[mqps] connect error: %s\n", mosquitto_strerror(rc));
        return false;
    }

    /* Run network loop until connected or timeout */
    struct timespec start, now;
    clock_gettime(CLOCK_MONOTONIC, &start);

    while (!ps->connected) {
        mosquitto_loop(ps->mosq, 10, 1);  /* 10ms loop */
        clock_gettime(CLOCK_MONOTONIC, &now);
        int elapsed_ms = (int)((now.tv_sec - start.tv_sec) * 1000 +
                               (now.tv_nsec - start.tv_nsec) / 1000000);
        if (elapsed_ms >= timeout_ms) {
            fprintf(stderr, "[mqps] connect timeout after %dms\n", elapsed_ms);
            return false;
        }
    }

    return true;
}

void mqps_disconnect(MqttPubSubStream *ps) {
    if (!ps || !ps->mosq) return;
    mosquitto_disconnect(ps->mosq);
    mosquitto_loop(ps->mosq, 100, 1);  /* flush */
    ps->connected = false;
}

bool mqps_is_connected(MqttPubSubStream *ps) {
    return ps && ps->connected;
}

bool mqps_publish(MqttPubSubStream *ps, const char *topic,
                  const void *payload, int payload_len,
                  int qos, bool retain) {
    if (!ps || !ps->mosq || !ps->connected) return false;

    int rc = mosquitto_publish(ps->mosq, NULL, topic, payload_len, payload,
                                qos, retain);
    if (rc == MOSQ_ERR_SUCCESS) {
        ps->published++;
        /* Flush to ensure publish goes out */
        mosquitto_loop(ps->mosq, 0, 1);
        return true;
    }

    fprintf(stderr, "[mqps] publish error: %s\n", mosquitto_strerror(rc));
    return false;
}

bool mqps_subscribe(MqttPubSubStream *ps, const char *topic, int qos) {
    if (!ps || !ps->mosq || !ps->connected) return false;

    int rc = mosquitto_subscribe(ps->mosq, NULL, topic, qos);
    if (rc == MOSQ_ERR_SUCCESS) {
        /* Run loop briefly to process SUBACK */
        mosquitto_loop(ps->mosq, 100, 1);
        return true;
    }

    fprintf(stderr, "[mqps] subscribe error: %s\n", mosquitto_strerror(rc));
    return false;
}

bool mqps_unsubscribe(MqttPubSubStream *ps, const char *topic) {
    if (!ps || !ps->mosq) return false;

    int rc = mosquitto_unsubscribe(ps->mosq, NULL, topic);
    if (rc == MOSQ_ERR_SUCCESS) {
        mosquitto_loop(ps->mosq, 100, 1);
        return true;
    }
    return false;
}

int mqps_poll(MqttPubSubStream *ps, int timeout_ms,
              MqpsMsgSlot *out_msgs, int max_msgs) {
    if (!ps || !ps->mosq) return 0;

    /* Run mosquitto network loop -- triggers on_message callbacks
       which push into the ring buffer. */
    struct timespec start, now;
    clock_gettime(CLOCK_MONOTONIC, &start);

    do {
        mosquitto_loop(ps->mosq, 1, 1);  /* 1ms per iteration */
        clock_gettime(CLOCK_MONOTONIC, &now);
        int elapsed_ms = (int)((now.tv_sec - start.tv_sec) * 1000 +
                               (now.tv_nsec - start.tv_nsec) / 1000000);
        if (elapsed_ms >= timeout_ms) break;
    } while (1);

    /* Drain ring buffer into output */
    int count = ring_pop(&ps->ring, out_msgs, max_msgs);
    ps->polled += (uint64_t)count;
    return count;
}

int mqps_drain(MqttPubSubStream *ps,
               MqpsMsgSlot *out_msgs, int max_msgs) {
    if (!ps) return 0;
    int count = ring_pop(&ps->ring, out_msgs, max_msgs);
    ps->polled += (uint64_t)count;
    return count;
}

void mqps_get_stats(MqttPubSubStream *ps, MqpsStats *out) {
    if (!ps || !out) return;
    out->published = ps->published;
    out->received  = ps->received;
    out->dropped   = atomic_load(&ps->ring.dropped);
    out->polled    = ps->polled;
    out->ring_used = ring_used(&ps->ring);
}

File: luajit_programs_and_containers/nano_data_center_instance/app_containers/mission_planner/container/planner/lib/lib/mqtt_pubsub.lua
--[[
    mqtt_pubsub.lua — LuaJIT FFI binding for streaming MQTT pub/sub.

    Subscribe once, messages buffer via C ring buffer, poll() drains them.
    No re-subscribe per read. Designed for low-latency robot bridges.

    Usage:
        local ps = PubSub.new("localhost", 1883, "my_client")
        ps:connect(5000)
        ps:subscribe("robots/rover_1/stream_bus", 1)

        -- In tick loop:
        local msgs = ps:poll(10)  -- 10ms timeout
        for _, m in ipairs(msgs) do
            print(m.topic, m.payload)
        end

        ps:publish("robots/rover_1/rpc", '{"cmd":"go"}', 1, false)
        ps:disconnect()
        ps:destroy()
]]

local ffi = require("ffi")

ffi.cdef[[
    /* Message slot (fixed-size, ring buffer element) */
    typedef struct {
        char     topic[256];
        char     payload[8192];
        uint32_t payload_len;
        int      qos;
        bool     retain;
        bool     valid;
    } MqpsMsgSlot;

    /* Stats */
    typedef struct {
        uint64_t published;
        uint64_t received;
        uint64_t dropped;
        uint64_t polled;
        uint32_t ring_used;
    } MqpsStats;

    /* Opaque handle */
    typedef struct MqttPubSubStream MqttPubSubStream;

    MqttPubSubStream *mqps_create(const char *host, int port,
                                   const char *client_id);
    void              mqps_destroy(MqttPubSubStream *ps);

    bool mqps_connect(MqttPubSubStream *ps, int timeout_ms);
    void mqps_disconnect(MqttPubSubStream *ps);
    bool mqps_is_connected(MqttPubSubStream *ps);

    bool mqps_publish(MqttPubSubStream *ps, const char *topic,
                      const void *payload, int payload_len,
                      int qos, bool retain);

    bool mqps_subscribe(MqttPubSubStream *ps, const char *topic, int qos);
    bool mqps_unsubscribe(MqttPubSubStream *ps, const char *topic);

    int mqps_poll(MqttPubSubStream *ps, int timeout_ms,
                  MqpsMsgSlot *out_msgs, int max_msgs);

    int mqps_drain(MqttPubSubStream *ps,
                   MqpsMsgSlot *out_msgs, int max_msgs);

    void mqps_get_stats(MqttPubSubStream *ps, MqpsStats *out);
]]

local C = ffi.load("mqtt_pubsub")

-- Pre-allocate poll buffer (reused across calls)
local MAX_POLL = 64
local poll_buf = ffi.new("MqpsMsgSlot[?]", MAX_POLL)

local PubSub = {}
PubSub.__index = PubSub

function PubSub.new(host, port, client_id)
    local h = C.mqps_create(host or "localhost", port or 1883, client_id)
    if h == nil then error("Failed to create MQTT PubSub", 2) end
    return setmetatable({ _h = h }, PubSub)
end

function PubSub:destroy()
    if self._h then
        C.mqps_destroy(self._h)
        self._h = nil
    end
end

function PubSub:connect(timeout_ms)
    if not C.mqps_connect(self._h, timeout_ms or 5000) then
        error("MQTT PubSub connect failed", 2)
    end
end

function PubSub:disconnect()
    C.mqps_disconnect(self._h)
end

function PubSub:is_connected()
    return C.mqps_is_connected(self._h)
end

function PubSub:publish(topic, payload, qos, retain)
    local len = payload and #payload or 0
    return C.mqps_publish(self._h, topic, payload, len, qos or 1, retain or false)
end

function PubSub:subscribe(topic, qos)
    if not C.mqps_subscribe(self._h, topic, qos or 1) then
        error("MQTT PubSub subscribe failed: " .. topic, 2)
    end
end

function PubSub:unsubscribe(topic)
    return C.mqps_unsubscribe(self._h, topic)
end

--- Poll for messages. Runs mosquitto loop for timeout_ms, then drains buffer.
--- Returns a Lua table of { topic=string, payload=string, qos=int, retain=bool }.
function PubSub:poll(timeout_ms, max_msgs)
    max_msgs = max_msgs or MAX_POLL
    if max_msgs > MAX_POLL then max_msgs = MAX_POLL end

    local count = C.mqps_poll(self._h, timeout_ms or 10, poll_buf, max_msgs)
    local result = {}
    for i = 0, count - 1 do
        local slot = poll_buf[i]
        result[#result + 1] = {
            topic   = ffi.string(slot.topic),
            payload = ffi.string(slot.payload, slot.payload_len),
            qos     = slot.qos,
            retain  = slot.retain,
        }
    end
    return result
end

--- Non-blocking drain: return already-buffered messages without waiting.
function PubSub:drain(max_msgs)
    max_msgs = max_msgs or MAX_POLL
    if max_msgs > MAX_POLL then max_msgs = MAX_POLL end

    local count = C.mqps_drain(self._h, poll_buf, max_msgs)
    local result = {}
    for i = 0, count - 1 do
        local slot = poll_buf[i]
        result[#result + 1] = {
            topic   = ffi.string(slot.topic),
            payload = ffi.string(slot.payload, slot.payload_len),
            qos     = slot.qos,
            retain  = slot.retain,
        }
    end
    return result
end

--- Get stats: { published, received, dropped, polled, ring_used }
function PubSub:get_stats()
    local s = ffi.new("MqpsStats")
    C.mqps_get_stats(self._h, s)
    return {
        published = tonumber(s.published),
        received  = tonumber(s.received),
        dropped   = tonumber(s.dropped),
        polled    = tonumber(s.polled),
        ring_used = tonumber(s.ring_used),
    }
end

return { PubSub = PubSub, _C = C }
